Mappings: Validate dynamic mappings updates on the master node.

jpountz opened the pull request:
This commit changes dynamic mappings updates so that they are synchronous on the
entire cluster and their validity is checked by the master node. There are some
important consequences of this commit:
 - a failing index request on a non-existing type does not implicitely create
   the type anymore
 - dynamic mappings updates cannot create inconsistent mappings on different
   shards
 - indexing requests that introduce new fields might induce latency spikes
   because of the overhead to update the mappings on the master node

Close #8688

jpountz commented:
This is still a work-in-progress as I have test failures on RiverTests. I am suspecting this is because this test triggers lots of concurrent dynamic mappings updates and there is a deadlock somehow because of my ignorange of how our transport works (eg. it is ok to update the mappings like I do in MapperUtils.validateDynamicMappingsUpdateOnMaster?). Any help would be greatly appreciated.

rjernst commented on src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java:
```diff
@@ -157,34 +155,12 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
+            try {
+                MapperUtils.validateDynamicMappingsUpdateOnMaster(client, logger, indexService.index().name(), mappingUpdate.getKey(), mappingUpdate.getValue(), waitForMappingUpdatePostRecovery);
+            } catch (Throwable t) {
+                logger.debug("failed to send mapping update post recovery to master for [{}]", t, mappingUpdate.getKey());
```
This seems like a bad case to get into? At least should it be warn or error level? This means we reapplied some updates from the translog, but the master rejected those updates...but we dont seem to do anything with the validation, so does that mean the mappings have already been updated in place? Also, could we limit the catch to just whatever exception would mean failed validation? Otherwise a bug as simple as an NPE in validate would get caught and just logged?

rjernst commented on src/main/java/org/elasticsearch/index/mapper/MapperUtils.java:
```diff
@@ -76,7 +82,50 @@ public void addConflict(String mergeFailure) {
+     * Merge {@code mergeWith} into {@code mergeTo}. Note: this method only
+     * merges mappings, not lookup structures. Conflicts are returned as exceptions.
+     */
+    public static void merge(Mapping mergeInto, Mapping mergeWith) {
```
This seems like a duplicate of the method above?

jpountz replied:
It does the same thing indeed, but on a Mapping object instead of a Mapper.

rjernst commented on src/main/java/org/elasticsearch/index/mapper/MapperUtils.java:
```diff
@@ -76,7 +82,50 @@ public void addConflict(String mergeFailure) {
+        }
+        final PutMappingResponse response = requestBuilder.get();
+        if (response.isAcknowledged() == false) {
+            throw new ElasticsearchIllegalStateException("Failed to validate dynamic mappings updates on the master node - response was not acknowledged");
```
Is illegal state the right exception to use? I would normally use this for inconsistent state (meaning we have broken code)?

rjernst commented on src/main/java/org/elasticsearch/index/mapper/Mapping.java:
```diff
@@ -0,0 +1,147 @@
+            }
+        }
+
+        if (!mergeContext.mergeFlags().simulate()) {
```
Use `== false` like we do in many other places?

rjernst commented on src/main/java/org/elasticsearch/percolator/PercolatorService.java:
```diff
@@ -280,10 +282,13 @@ private ParsedDocument parseRequest(IndexService documentIndexService, Percolate
+                        if (docMapper.v2() != null) {
+                            doc.addDynamicMappingsUpdate(docMapper.v2());
+                        }
+                        if (doc.dynamicMappingsUpdate() != null) {
```
could you not merge this with the `if` above?

jpountz replied:
I don't think we can? It might happen that the first condition is false and the second is true if you only got dynamic updates through `copy_to` directives (which needs to be handled differently since it can insert dynamic mappings at arbitrary places in the mappings).

rjernst commented on src/test/java/org/elasticsearch/index/mapper/update/default_mapping_with_disabled_root_types.json:
```diff
@@ -1 +1 @@
-{"type":{"_timestamp":{"enabled":false},"_index":{"enabled":false},"_size":{"enabled":false},"properties":{}}}
\ No newline at end of file
+{"type":{"_timestamp":{"enabled":false},"_index":{"enabled":false},"_size":{"enabled":false}}}
```
removed newline?

jpountz replied:
Yes, so that tests can check `"foo".equals(string)` instead of `"foo\n".equals(string)`

rjernst commented:
The mappings side looks great! My main question is about what to do on translog replay if sending mapping updates to the master fails.

jpountz commented:
Good question. This should hopefully not happen once mappings updates are synchronous but I believe this could happen when upgrading from 1.x. In that case I don't think we can do better than warning a scary message to the logs?

rjernst commented:
@jpountz Ok, let's put it at warn level then?

bleskes commented:
I think we should fail the shard in this case. It means that some data was indexed in a way that is inconsistent with the “master” mapping in which case you probably have rogue lucene indices. Picking up on another copy is a better alternative. If we ending up having no copies we are at least clearly communicating things are bad (master have a different mapping that is inconsistent with any shard). If people want to still recover they can intervene and delete the translog (which may loose some data but I think it’s an acceptable solution for this bad place to be in). Also note that we flush on close during node shutdown, so the chance this will happen is small...


> On 19 Apr 2015, at 04:01, Ryan Ernst <notifications@github.com> wrote:
> 
> @jpountz Ok, let's put it at warn level then?
> 
> —
> Reply to this email directly or view it on GitHub.
>

jpountz commented:
I added (and beasted) more tests about concurrent indexing requests and the rivers issue seems to be specific to rivers actually so I left mapping updates handled like today when it comes to rivers (ie. updated locally first and then propagated to the master).

Also failing to apply a mapping update on recovery now fails the shard.

kimchy commented on src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java:
```diff
@@ -151,41 +151,49 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
-                            logger.debug("waited for mapping update on master for [{}], yet timed out", type);
+    private void validateMappingUpdate(final String type, Mapping update) {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final Throwable[] error = new Throwable[1];
```
can this be an AtomicReference?

bleskes commented on src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java:
```diff
@@ -151,41 +151,49 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
+            @Override
+            public void run() throws InterruptedException {
+                try {
+                    if (latch.await(waitForMappingUpdatePostRecovery.millis(), TimeUnit.MILLISECONDS) == false) {
```
waitForMappingUpdatePostRecovery defaults to 30s, give that the timeout now results in a failed shard (good!) I think we should be more lenient. Especially given the fact that local gateway recovery runs on full cluster restart where the master might be overloaded by things to do. How about something very conservative like 15m (which is what we use for the same update mapping  - see RecoverySourceHandler#updateMappingOnMaster)

jpountz commented:
@bleskes @kimchy Pushed a new commit to address your comments

purduemike commented:
+1 this is happening everyday for me during an index rollover.
Looks like this fix is slated for v2.0.0.  Can we instead merged into the next minor release?
